hooks/pre-checkout

```diff
@@ -3,56 +3,123 @@ set -eo pipefail
+s3_download () {
+  # Fetch the most recent S3 backup key
+	SNAPSHOT_KEY=$(aws --output json s3api list-objects-v2 --bucket "$BUCKET" --prefix "$REPO_PATH" --query 'Contents[].Key' | jq -sr 'sort[] | reverse[] | select(endswith(".git.tar"))' | head -n1 )
 
-	# Exit early if there's no available snapshot
+  # Exit early if there's no available snapshot
 	if [ -z "$SNAPSHOT_KEY" ]; then
-		echo "No snapshots found in $BUCKET"
-		exit 0
+		echo "  ↳ WARNING: No snapshots found in $BUCKET"
+    ERRORS_ARRAY+=("- WARNING: No snapshots found in $BUCKET")
+		return 1
 	fi
 
-	echo "Downloading snapshot: $SNAPSHOT_KEY to $DESTINATION"
+ SNAPSHOT_KEY_CHECK=$(echo "$SNAPSHOT_KEY" | cut -d "/" -f 1-2)
+  if [ "$SNAPSHOT_KEY_CHECK" == "$REPO_PATH" ]; then
+		echo "Downloading snapshot: $SNAPSHOT_KEY to $DESTINATION"
+    # Then download it
+    rm -rf "$DESTINATION" # Delete before starting, just in case
+    aws s3 cp "s3://$BUCKET/$SNAPSHOT_KEY" "$DESTINATION"
+    echo ":large_green_circle: S3 Download Succeeded"
+    return 0
+  else
+    echo "  ↳ WARNING: Nothing downloaded. $SNAPSHOT_KEY is not the same as $REPO_PATH (comparison used $SNAPSHOT_KEY_CHECK)"
+    ERRORS_ARRAY+=("- WARNING: Nothing downloaded. $SNAPSHOT_KEY is not the same as $REPO_PATH")
+    return 1
+  fi
+}
+
+annotate_error() {
+  ERRORS_ARRAY=("$@")
+  ERRORS=$(
+    for error in "${ERRORS_ARRAY[@]}"
+    do
+      echo "$error"
+    done
+  )
 
-	# Then download it
-	rm -rf $DESTINATION # Delete before starting, just in case
-	aws s3 cp "s3://$BUCKET/$SNAPSHOT_KEY" "$DESTINATION"
+  buildkite-agent annotate --style "warning" --context "warnings" --append \
+"
+Step **$BUILDKITE_LABEL** had _WARNINGS_!
+Agent: **$BUILDKITE_AGENT_NAME** <br>
+There were warnings in your build indicating the repo was expected to be cached, but wasn't found.<br>
+Please ensure you're repo is properly cached via the [bash-cache-buildkite-plugin](https://github.com/Automattic/bash-cache-buildkite-plugin).
+\`\`\`
+Errors:
+${ERRORS}
+\`\`\`
+"
+}
+
+###### BEGIN ######
+
+# If we have an available Git Mirror Server, use it
+if [ -n "${GIT_MIRROR_SERVER_ROOT:-}" ]; then
+  echo "Git mirror found, attempting download."
+  if ! git_mirror_download; then
+    echo "~~~ :warning: Git Mirror download failed, attempting download from s3."
+    ERRORS_ARRAY+=("- Git Mirror download failed, attempting download from s3.")
+    if ! s3_download; then
+      echo "~~~ :warning: S3 Download failed as well. Cache must not exist yet."
+      ERRORS_ARRAY+=("- S3 Download failed as well. Cache must not exist yet.")
+      if [ -n "${ERRORS_ARRAY[*]}" ]; then
+        annotate_error "${ERRORS_ARRAY[@]}"
+      fi
+      exit 0
+    fi
+  fi
+# Otherwise, use S3 directly
+else
+  if ! s3_download; then
+    echo "--- :warning: S3 Download failed. Cache must not exist yet."
```

Comment:
This isn't a given – `aws s3 cp` may fail for other reasons (local disk is full, AWS hiccups, network weather). Might be worth mentioning as a possibility but not a definitive.

Reply:
@jkmassel updated this to report more on s3 failures and fixed that strange spacing issue.